File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

vector<vector<int>> e;
vector<int>subtsize;

void dfs(int cur, int pre){
    subtsize[cur] = 1;
    for(auto nxt : e[cur]){
        if(nxt == pre) continue;
        dfs(nxt, cur);
        subtsize[cur] += subtsize[nxt];
    }
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, r, q; cin >> n >> r >> q; 
    e.resize(n+1); subtsize.assign(n+1, 0);
    for(int i = 0; i < n-1; i++){
        int u, v; cin >> u >> v;
        e[u].push_back(v);
        e[v].push_back(u);
    }
    dfs(r, 0);
    while(q--){
        int qry; cin >> qry;
        cout << subtsize[qry] << '\n';
    }
    return 0;
} 

File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

sys.setrecursionlimit(10**7)

def main():
    n, r, q = map(int, input().split())
    e = [[] for _ in range(n+1)]
    vis = [0]*(n+1)
    subtsize = [0]*(n+1)
    for _ in range(n-1):
        u, v = map(int, input().split())
        e[u].append(v)
        e[v].append(u)
    def dfs(cur):
        vis[cur] = subtsize[cur] = 1
        for nxt in e[cur]:
            if not vis[nxt]:
                dfs(nxt)
                subtsize[cur] += subtsize[nxt]                        
    dfs(r)
    ans = []
    for _ in range(q):
        qry = int(input())
        ans.append(subtsize[qry])
    print('\n'.join(map(str, ans)))
    return
if __name__ == "__main__":
    main()